libraries/c_sdk_v4/integration-test/core_mqtt_system_test.c

```diff
@@ -0,0 +1,1635 @@
+    {
+        /* Terminate TLS session and TCP connection to test session restoration
+         * across network connection. */
+        ( void ) SecureSocketsTransport_Disconnect( &networkContext );
+    }
+    else
+    {
+        /* Handle incoming publish. The lower 4 bits of the publish packet
+         * type is used for the dup, QoS, and retain flags. Hence masking
+         * out the lower bits to check if the packet is publish. */
+        if( ( pPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+        {
+            assert( pPublishInfo != NULL );
+            /* Handle incoming publish. */
+
+            /* Cache information about the incoming PUBLISH message to process
+             * in test case. */
+            memcpy( &incomingInfo, pPublishInfo, sizeof( MQTTPublishInfo_t ) );
+            incomingInfo.pTopicName = NULL;
+            incomingInfo.pPayload = NULL;
+            /* Allocate buffers and copy information of topic name and payload. */
+            incomingInfo.pTopicName = pvPortMalloc( pPublishInfo->topicNameLength );
+            TEST_ASSERT_NOT_NULL( incomingInfo.pTopicName );
+            memcpy( ( void * ) incomingInfo.pTopicName, pPublishInfo->pTopicName, pPublishInfo->topicNameLength );
+            incomingInfo.pPayload = pvPortMalloc( pPublishInfo->payloadLength );
+            TEST_ASSERT_NOT_NULL( incomingInfo.pPayload );
+            memcpy( ( void * ) incomingInfo.pPayload, pPublishInfo->pPayload, pPublishInfo->payloadLength );
+
+            /* Update the global variable if the incoming PUBLISH packet
+             * represents a retained message. */
+            receivedRetainedMessage = pPublishInfo->retain;
+        }
+        else
+        {
+            handleAckEvents( pPacketInfo, pDeserializedInfo->packetIdentifier );
+        }
+    }
+}
+
+static MQTTStatus_t subscribeToTopic( MQTTContext_t * pContext,
+                                      const char * pTopic,
+                                      MQTTQoS_t qos )
+{
+    MQTTSubscribeInfo_t pSubscriptionList[ 1 ];
+
+    assert( pContext != NULL );
+
+    /* Start with everything at 0. */
+    ( void ) memset( ( void * ) pSubscriptionList, 0x00, sizeof( pSubscriptionList ) );
+
+    pSubscriptionList[ 0 ].qos = qos;
+    pSubscriptionList[ 0 ].pTopicFilter = pTopic;
+    pSubscriptionList[ 0 ].topicFilterLength = strlen( pTopic );
+
+    /* Generate packet identifier for the SUBSCRIBE packet. */
+    globalSubscribePacketIdentifier = MQTT_GetPacketId( pContext );
+
+    /* Send SUBSCRIBE packet. */
+    return MQTT_Subscribe( pContext,
+                           pSubscriptionList,
+                           sizeof( pSubscriptionList ) / sizeof( MQTTSubscribeInfo_t ),
+                           globalSubscribePacketIdentifier );
+}
+
+static MQTTStatus_t unsubscribeFromTopic( MQTTContext_t * pContext,
+                                          const char * pTopic,
+                                          MQTTQoS_t qos )
```

Comment:
I don't think we need QoS for unsubscribe.

Reply:
Removed